open-craft/opencraft
# [BB-2239] Hero customization page

Replaces https://github.com/open-craft/opencraft/pull/553

Implements the changes for the hero customization page in the console and the corresponding changes in the backend.

**Testing instructions**:
* Checkout the PR branch.
* Apply the migrations and start the backend app server.
* Build the API client and start the frontend server.
* Create a new betatest instance using the new registration form.
* Verify the email addresses provided during the registration from the django shell by running the below statement for each distinct email address:
  ```python
  EmailAddress.objects.confirm(EmailAddress.objects.get(email='<email address>').key)
  ```
* Verify that the `draft_static_content_overrides` field has been populated with the below default value.
  ```python
  {
      "version": 0,
      "homepage_overlay_html": "<h1>Welcome to <instance name></h1><p>It works! Powered by  Open edX®</p>"
  }
  ```
* Also verify that the `home-page-hero-title-color` and `home-page-hero-subtitle-color` variables have been set to their default values in the `draft_theme_config` field.
* Log in to the user console from the new frontend.
* You will land on the `Instance settings -> General` section. If the fields in that page are missing, click on each field and click outside to trigger the `onblur` event to force the data to get loaded. 

  **Note**: This is an existing bug somewhere in the component lifecycle which affects multiple components including the components created in this PR. This will be investigated and fixed in a separate task and PR.
* Navigate to the `Theme -> Hero` section. Verify that the hero preview component shows up with the default hero title and subtitle.
* Also verify that the text boxes below for the title and subtitle fields are filled with the same values as shown in the preview.
* For all the steps below that involve editing, uploading, deleting, verify that the changes are stored as expected in the database.
* Type in those text boxes and verify that the preview is updated when the `onblur` event happens.
* Delete all the text in either of the text boxes and click outside. Verify that the default value gets filled in that case.
* Verify that the color input fields corresponding to the title and subtitle colors have the correct value set and that changing those values updates the color in the preview.
* Upload a cover image of appropriate dimensions for the hero section and verify that it is set as the background image for the hero section, replacing the default shade of grey. Also verify that a `remove` link shows up below the upload button once the image is successfully uploaded.
* Click the `remove` link and verify that the background image is removed.
* Try refreshing the page. If the fields become empty. Visit the `Instance settings -> General` section, confirm that the fields there have the values filled and return to the Hero customization page.
* Verify that the preview component shows the current stored state as expected and all the fields have the expected values.
* Perform any other tests of your choice to try and make the component break in unexpected ways.
* No breakages for existing, un-migrated instances.
  * On the `new-console` branch create a new instance from the old registration form.
  * Switch to the PR branch and run the app.
  * Log in using the account created above via the new frontend.
  * Verify that the hero customization page loads without any errors but doesn't have all its functionality exposed.

**Author notes and concerns**:
* The preview of the background image will not be a 100% accurate and match what will be seen on the instance. This is because the hero preview section is a lot smaller in size than the LMS hero section and the background image there is set with the `background-size` property set to `cover`, which will stretch the image to fit the box. In this PR, I have scaled the uploaded image to fit the dimensions of the box that is there in the preview component. There is no way to give a 100% accurate result without having the dimensions of the preview component match that of the LMS.
* The component lifecycle issue affecting the components added in this PR and other existing components hasn't been fixed in this PR.
* The data migration to migrate the existing instances to pre-populate the default values for the `draft_static_content_overrides` field which are needed for the changes in this PR hasn't been created in this PR. This is because the changes in this PR also require migrating the instance theme configuration to the v1 version and that hasn't been done yet.
